homeassistant/components/stream/__init__.py

```diff
@@ -95,6 +100,80 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class StreamClientError(StrEnum):
+    """Enum for stream client errors."""
+
+    BadRequest = auto()
+    Unauthorized = auto()
+    Forbidden = auto()
+    NotFound = auto()
+    Other = auto()
+
+
+class StreamOpenClientError(HomeAssistantError):
+    """Raised when client error received when trying to open a stream.
+
+    :param stream_client_error: The type of client error
+    """
+
+    def __init__(
+        self, *args: Any, stream_client_error: StreamClientError, **kwargs: Any
+    ) -> None:
+        self.stream_client_error = stream_client_error
+        super().__init__(*args, **kwargs)
+
+
+async def _async_try_open_stream(
+    hass: HomeAssistant, source: str, pyav_options: dict[str, str] | None = None
+) -> InputContainer | OutputContainer:
+    """Try to open a stream.
+
+    Will raise StreamOpenClientError if an http client error is encountered.
+    """
+    import av  # pylint: disable=import-outside-toplevel
+
+    if not pyav_options:
+        pyav_options = {}
+
+    default_pyav_options = {
+        "rtsp_flags": CONF_PREFER_TCP,
+        "timeout": str(SOURCE_TIMEOUT),
+    }
+
+    pyav_options = {
+        **default_pyav_options,
+        **pyav_options,
+    }
+    error_lookup = {
+        av.HTTPBadRequestError: StreamClientError.BadRequest,
+        av.HTTPUnauthorizedError: StreamClientError.Unauthorized,
+        av.HTTPForbiddenError: StreamClientError.Forbidden,
+        av.HTTPNotFoundError: StreamClientError.NotFound,
+        av.HTTPOtherClientError: StreamClientError.Other,
+    }
+
+    try:
+        func = partial(av.open, source, options=pyav_options, timeout=5)
+        container = await hass.loop.run_in_executor(None, func)
+    except av.HTTPClientError as ex:
+        client_error = error_lookup.get(ex.__class__, StreamClientError.Other)
+        raise StreamOpenClientError(stream_client_error=client_error) from ex
```

Comment:
I think here i might prefer just using the built in exception handing hierarchy to catch rather than looking up the exception type in a dict.

Reply:
That's done